src/ioctls/vm.rs

```diff
@@ -1177,18 +1177,18 @@ impl VmFd {
     /// };
     /// // On ARM, creating VGICv3 may fail due to hardware dependency.
     /// // Retry to create VGICv2 in that case.
-    /// let device_fd = vm.create_device(&mut device).unwrap_or_else(|_| {
+    /// let device_fd = vm.create_device(&device).unwrap_or_else(|_| {
     ///     #[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
     ///     panic!("Cannot create VFIO device.");
     ///     #[cfg(any(target_arch = "arm", target_arch = "aarch64"))]
     ///     {
     ///         device.type_ = kvm_device_type_KVM_DEV_TYPE_ARM_VGIC_V2;
-    ///         vm.create_device(&mut device)
+    ///         vm.create_device(&device)
     ///             .expect("Cannot create vGIC device")
     ///     }
     /// });
     /// ```
-    pub fn create_device(&self, device: &mut kvm_create_device) -> Result<DeviceFd> {
+    pub fn create_device(&self, device: &kvm_create_device) -> Result<DeviceFd> {
```

Comment:
Mh, maybe I'm missing some context here, but wouldnt the below be a better signature for this function?

```suggestion
    pub fn create_device(&mut self, device: kvm_create_device) -> Result<DeviceFd> {
```

We take `self` by mutable reference, because we're modifying the virtual machine that `self` represents, and we take ownership of the `kvm_create_device` object, because in the success case we return ownership over a `File` created from the `fd` inside the `kvm_create_device`, so we "transfer" ownership here

Reply:
Ah, sorry, I did indeed misunderstand what is going on here. The `KVM_CREATE_DEVICE` ioctl writes to the `fd` field of the `kvm_create_device` structure, so that should indeed be passed by mutable reference (since it's getting mutated). The error in the code is that it uses `ioctl_with_ref` instead of `ioctl_with_mut_ref`.

Reply:
So if I'm understanding correctly, instead of modifying `kvm_create_device` to be a immutable reference, the actual change is that it should call `ioctl_with_mut_ref` (since the ioctl is what will change data in `kvm_create_device`).

Is that accurate? This change was submitted because `clippy` was complaining about the mutable reference. I was missing the underlying context that `ioctl_with_mut_ref` should be called.

Reply:
Yep, that's right :)
There actually seem to be a few more instances of this (e.g. where `ioctl_with_ref` is used for an ioctl that actually mutates its input), we should fix those too (although we can do that outside of this PR if you want)

Reply:
I can close this PR, and will submit some changes on another. Thanks for clearing this up.